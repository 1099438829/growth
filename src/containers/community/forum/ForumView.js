import React, { Component, PropTypes } from 'react';
import {
  View,
  ListView,
  RefreshControl,
  Text,
} from 'react-native';
import InfiniteScrollView from 'react-native-infinite-scroll-view';

import AppColors from '../../../theme/colors';

class ForumView extends Component {
  static componentName = 'ForumView';

  static propTypes = {
    data: PropTypes.arrayOf(PropTypes.object).isRequired,
    canLoadMoreContent: PropTypes.bool.isRequired,
    onLoadMoreAsync: PropTypes.func.isRequired,
    element: PropTypes.func.isRequired,
    reFetch: PropTypes.func.isRequired,
    dataUrl: PropTypes.string.isRequired,
  };

  constructor() {
    super();

    this.state = {
      isRefreshing: true,
      canLoadMoreContent: false,
      isLoadMoreAsync: null,
      dataSource: new ListView.DataSource({
        rowHasChanged: (row1, row2) => row1 !== row2,
      }),
    };
  }

  componentWillMount() {
    // TODO: check componentWillReceiveProps problem
    if (this.props.data.length > 0 && this.state.dataSource.getRowCount() < 1) {
      this.setState({
        dataSource: this.getUpdatedDataSource(this.props),
        canLoadMoreContent: this.props.canLoadMoreContent,
        dataUrl: this.props.dataUrl,
        isRefreshing: false,
      });
    }
  }

  componentWillReceiveProps(props) {
    this.setState({
      dataSource: this.getUpdatedDataSource(props),
      canLoadMoreContent: props.canLoadMoreContent,
      dataUrl: props.dataUrl,
      isRefreshing: false,
    });
  }

  onLoadMoreAsync = () => {
    this.setState({ isLoadMoreAsync: true });
    this.props.onLoadMoreAsync()
    .then(() => {
      this.setState({ isLoadMoreAsync: false });
    });
  };

  getUpdatedDataSource(props) {
    const rows = props.data;
    const ids = rows.map((obj, index) => index);

    return this.state.dataSource.cloneWithRows(rows, ids);
  }

  reFetch = () => {
    if (this.props.reFetch) {
      this.setState({ isRefreshing: true });

      this.props.reFetch({ dataUrl: this.state.dataUrl })
      .then(() => {
        this.setState({ isRefreshing: false });
      });
    }
  };


  render = () => {
    const { data } = this.props;
    const { isRefreshing, dataSource, canLoadMoreContent } = this.state;

    if (!isRefreshing && (!data || data.length < 1)) {
      return <Text>Error</Text>;
    }

    return (
      <View>
        <ListView
          initialListSize={10}
          renderScrollComponent={props => <InfiniteScrollView {...props} />}
          renderRow={this.props.element}
          dataSource={dataSource}
          canLoadMore={canLoadMoreContent}
          onLoadMoreAsync={this.onLoadMoreAsync}
          refreshControl={
            this.props.reFetch ?
              <RefreshControl
                refreshing={isRefreshing}
                onRefresh={this.reFetch}
                tintColor={AppColors.brand.primary}
              />
              : null
          }
        />

      </View>
    );
  };
}

export default ForumView;
