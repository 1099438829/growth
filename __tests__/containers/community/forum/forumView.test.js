import React from 'react';
import renderer from 'react-test-renderer';
import ForumView from '../../../../src/containers/community/forum/ForumView';

describe('test user center', () => {
  it('renders correctly', () => {
    const tree = renderer.create(
      <ForumView
        data={[]}
        canLoadMoreContent={false}
        onLoadMoreAsync={() => {}}
        reFetch={() => {}}
        dataUrl={'https://forum.growth.ren/api/discussions'}
        element={() => {}}
      />,
    );

    const treeJson = tree.toJSON();
    expect(treeJson).toMatchSnapshot();
  });
});
