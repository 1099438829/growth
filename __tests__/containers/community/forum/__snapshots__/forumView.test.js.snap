// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test forum render renders correctly 1`] = `<View />`;
